File: src/main/java/com/cfgglobal/test/service/base/BaseService.java
package com.cfgglobal.test.service.base;

import com.cfgglobal.test.base.Filter;
import com.cfgglobal.test.dao.base.BaseDao;
import com.cfgglobal.test.service.rule.SecurityFilter;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import org.joor.Reflect;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.data.repository.NoRepositoryBean;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.transaction.annotation.Transactional;

import java.io.Serializable;


@NoRepositoryBean
public abstract class BaseService<T, ID extends Serializable> {


    protected BaseDao<T, ID> baseDao;

    @Autowired
    protected SecurityFilter securityFilter;

    @Autowired
    public void setBaseDao(BaseDao<T, ID> baseDao) {
        this.baseDao = baseDao;
    }

    public Page<T> findByRequestParameters(Map<String, String[]> info, Pageable pageable) {
        return baseDao.findByRequestParameters(info, pageable);
    }


    public List<T> findByRequestParameters(Map<String, String[]> info) {
        return baseDao.findByRequestParameters(info);
    }


    public Page<T> findByFilter(List<Filter> filters, Pageable pageable) {
        return baseDao.findByFilter(filters, pageable);
    }


    public List<T> findByFilter(List<Filter> filters) {
        return baseDao.findByFilter(filters);
    }

    public List<T> findByFilter(Filter filter) {
        return baseDao.findByFilter(List.of(filter));
    }


    @Transactional
    public Page<T> findBySecurity(String method, String requestURI, Map<String, String[]> params, Pageable pageable) {
        //   ConfigAttribute configAttribute = MyInvocationSecurityMetadataSourceService.getConfigAttributeDefinition();
        //  System.out.println(configAttribute);
        List<Filter> queryFilters = Filter.createFilters(params);
        List<Filter> securityFilters = securityFilter.query(method, requestURI);
        return baseDao.findByFilter(queryFilters.appendAll(securityFilters), pageable);
    }

    @Transactional
    public List<T> findBySecurity(String method, String requestURI, Map<String, String[]> params) {
        //   ConfigAttribute configAttribute = MyInvocationSecurityMetadataSourceService.getConfigAttributeDefinition();
        //  System.out.println(configAttribute);
        List<Filter> queryFilters = Filter.createFilters(params);
        List<Filter> securityFilters = securityFilter.query(method, requestURI);
        return baseDao.findByFilter(queryFilters.appendAll(securityFilters));
    }

    public T findOneBySecurity(Long id, String method, String requestURI) {
        List<Filter> securityFilters = securityFilter.query(method, requestURI);
        List<T> list = baseDao.findByFilter(securityFilters);
        T entity = baseDao.findById((ID) id).get();
        if (list.contains(entity)) {
            return entity;
        } else {
            throw new AccessDeniedException(requestURI);
        }
    }

    public void deleteBySecurity(Long id, String method, String requestURI) {
        List<Filter> securityFilters = securityFilter.query(method, requestURI);
        List<T> list = baseDao.findByFilter(securityFilters);
        T entity = baseDao.findById((ID) id).get();
        if (list.contains(entity)) {
            baseDao.delete(entity);
        } else {
            throw new AccessDeniedException(requestURI);
        }
    }

    public <S extends T> S saveBySecurity(S entity, String method, String requestURI) {
        Object id = Reflect.on(entity).get("id");
        if (id == null || baseDao.findByFilter(securityFilter.query(method, requestURI)).contains(entity)) {
            return save(entity);
        }
        throw new AccessDeniedException(requestURI);
    }


    public Page<T> findAll(Pageable pageable) {
        return baseDao.findAll(pageable);
    }

    public List<T> findAll() {
        return List.ofAll(baseDao.findAll());
    }

    public List<T> findAll(Sort sort) {
        return List.ofAll(baseDao.findAll(sort));
    }

    public List<T> findAll(Iterable<ID> ids) {
        return List.ofAll(baseDao.findAll(ids));
    }

    public long count() {
        return baseDao.count();
    }

    public void delete(ID id) {
        baseDao.delete(id);
    }

    public void delete(T entity) {
        baseDao.delete(entity);
    }


    public void deleteAll() {
        baseDao.deleteAll();
    }

    public <S extends T> List<S> save(Iterable<S> entities) {
        return List.ofAll(baseDao.save(entities));
    }

    public void flush() {
        baseDao.flush();
    }

    public <S extends T> S saveAndFlush(S entity) {
        return baseDao.saveAndFlush(entity);
    }

    public void deleteInBatch(Iterable<T> entities) {
        baseDao.deleteInBatch(entities);
    }

    public void deleteAllInBatch() {
        baseDao.deleteAllInBatch();
    }

    public <S extends T> List<S> findAll(Example<S> example) {
        return List.ofAll(baseDao.findAll(example));
    }

    public <S extends T> List<S> findAll(Example<S> example, Sort sort) {
        return List.ofAll(baseDao.findAll(example, sort));
    }


    public List<T> findAll(Specification<T> spec) {
        return List.ofAll(baseDao.findAll(spec));
    }

    public Page<T> findAll(Specification<T> spec, Pageable pageable) {
        return baseDao.findAll(spec, pageable);
    }


    public List<T> findAll(Specification<T> spec, Sort sort) {
        return List.ofAll(baseDao.findAll(spec, sort));
    }

    public long count(Specification<T> spec) {
        return baseDao.count(spec);
    }

    public <S extends T> S save(S entity) {
        return baseDao.save(entity);
    }

    public T findOne(ID id) {
        return baseDao.findById(id).get();
    }

}

File: src/main/kotlin/com/cfgglobal/test/service/base/BaseService.kt
package com.cfgglobal.test.service.base


import arrow.core.None
import arrow.core.Option
import arrow.core.Some
import com.cfgglobal.test.dao.base.BaseDao
import com.cfgglobal.test.service.rule.SecurityFilter
import com.cfgglobal.test.vo.Filter
import org.joor.Reflect
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Example
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.domain.Specification
import org.springframework.security.access.AccessDeniedException
import org.springframework.stereotype.Component
import java.io.Serializable

@Component
abstract class BaseService<T, ID : Serializable> {
    @Autowired
    lateinit var baseDao: BaseDao<T, ID>

    @Autowired
    lateinit var securityFilter: SecurityFilter


    fun findByRequestParameters(info: Map<String, Array<String>>, pageable: Pageable): Page<T> {
        return baseDao.findByRequestParameters(info, pageable)
    }


    fun findByRequestParameters(info: Map<String, Array<String>>): List<T> {
        return baseDao.findByRequestParameters(info)
    }


    fun findByFilter(filters: List<Filter>, pageable: Pageable): Page<T> {
        return baseDao.findByFilter(filters, pageable)
    }


    fun findByFilter(filters: List<Filter>): List<T> {
        return baseDao.findByFilter(filters)
    }

    fun findByFilter(filter: Filter): List<T> {
        return baseDao.findByFilter(listOf(filter))
    }


    fun findBySecurity(method: String, requestURI: String, params: Map<String, Array<String>>, pageable: Pageable): Page<T> {
        //  ConfigAttribute configAttribute = MyInvocationSecurityMetadataSourceService.getConfigAttributeDefinition();
        //  System.out.println(configAttribute);
        val queryFilters = Filter.createFilters(params)
        val securityFilters = securityFilter.query(method, requestURI)
        return baseDao.findByFilter(queryFilters + securityFilters, pageable)
    }

    fun findBySecurity(method: String, requestURI: String, params: Map<String, Array<String>>): List<T> {
        //   ConfigAttribute configAttribute = MyInvocationSecurityMetadataSourceService.getConfigAttributeDefinition();
        //  System.out.println(configAttribute);
        val queryFilters = Filter.createFilters(params)
        val securityFilters = securityFilter.query(method, requestURI)
        return baseDao.findByFilter(queryFilters + securityFilters)
    }

    fun findOneBySecurity(id: ID, method: String, requestURI: String): T {
        val securityFilters = securityFilter.query(method, requestURI)
        val list = baseDao.findByFilter(securityFilters)
        val entity = findOne(id)
        return when (entity) {
            is Some -> {
                if (list.contains(entity.t)) {
                    entity.t
                } else {
                    throw AccessDeniedException(requestURI)
                }
            }
            None -> throw AccessDeniedException(requestURI)
        }
    }

    fun deleteBySecurity(id: ID, method: String, requestURI: String) {
        val securityFilters = securityFilter.query(method, requestURI)
        val list = baseDao.findByFilter(securityFilters)
        val entity = findOne(id)
        when (entity) {
            is Some -> {
                if (list.contains(entity.t)) {
                    baseDao.delete(entity.t)
                } else {
                    throw AccessDeniedException(requestURI)
                }
            }
            None -> throw AccessDeniedException(requestURI)
        }

    }

    fun <S : T> saveBySecurity(entity: S, method: String, requestURI: String): S {
        val id = Reflect.on(entity).get<Any>("id")
        if (id == null || baseDao.findByFilter(securityFilter!!.query(method, requestURI)).contains(entity)) {
            return save(entity)
        }
        throw AccessDeniedException(requestURI)
    }


    fun findAll(pageable: Pageable): Page<T> {
        return baseDao.findAll(pageable)
    }

    fun findAll(): List<T> {
        return baseDao.findAll()
    }

    fun findAll(sort: Sort): List<T> {
        return baseDao.findAll(sort)
    }

    fun findAll(ids: Iterable<ID>): List<T> {
        return baseDao.findAllById(ids)
    }

    fun count(): Long {
        return baseDao.count()
    }

    fun delete(id: ID) {
        baseDao.deleteById(id)
    }

    fun delete(entity: T) {
        baseDao.delete(entity)
    }


    fun deleteAll() {
        baseDao.deleteAll()
    }

    fun <S : T> saveAll(entities: Iterable<S>): List<S> {
        return baseDao.saveAll(entities)
    }

    fun flush() {
        baseDao.flush()
    }

    fun <S : T> saveAndFlush(entity: S): S {
        return baseDao.saveAndFlush(entity)
    }

    fun deleteInBatch(entities: Iterable<T>) {
        baseDao.deleteInBatch(entities)
    }

    fun deleteAllInBatch() {
        baseDao.deleteAllInBatch()
    }

    fun <S : T> findAll(example: Example<S>): List<S> {
        return baseDao.findAll(example)
    }

    fun <S : T> findAll(example: Example<S>, sort: Sort): List<S> {
        return baseDao.findAll(example, sort)
    }


    fun findAll(spec: Specification<T>): List<T> {
        return baseDao.findAll(spec)
    }

    fun findAll(spec: Specification<T>, pageable: Pageable): Page<T> {
        return baseDao.findAll(spec, pageable)
    }


    fun findAll(spec: Specification<T>, sort: Sort): List<T> {
        return baseDao.findAll(spec, sort)
    }

    fun count(spec: Specification<T>): Long {
        return baseDao.count(spec)
    }

    fun <S : T> save(entity: S): S {
        return baseDao.save(entity)
    }

    fun findOne(id: ID): Option<T> {
        return baseDao.findById(id)
                .map { Option.fromNullable(it) }
                .orElse(Option.empty())

    }

}


